ownership-chain/pallets/asset-metadata-extender/src/lib.rs

```diff
@@ -0,0 +1,126 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub mod traits;
+pub mod types;
+
+use frame_support::pallet_prelude::*;
+pub use pallet::*;
+use sp_runtime::{traits::One, ArithmeticError, DispatchResult};
+use traits::AssetMetadataExtender;
+use types::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	/// Configure the pallet by specifying the parameters and types on which it depends.
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// Because this pallet emits events, it depends on the runtime's definition of an event.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Limit for the length of `token_uri`
+		#[pallet::constant]
+		type MaxTokenUriLength: Get<u32>;
+
+		/// Limit for the length of `universal_location`
+		#[pallet::constant]
+		type MaxUniversalLocationLength: Get<u32>;
+	}
+
+	/// Asset metadata extensions counter for a given location
+	#[pallet::storage]
+	#[pallet::getter(fn metadata_extensions_counter)]
+	pub(super) type MetadataExtensionsCounter<T: Config> =
+		StorageMap<_, Blake2_128Concat, UniversalLocationOf<T>, Index, ValueQuery>;
+
+	/// Records all metadata extensions with index for a given asset location
+	#[pallet::storage]
+	#[pallet::getter(fn indexed_metadata_extensions_details)]
+	pub(super) type IndexedMetadataExtensionsDetails<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		UniversalLocationOf<T>,
+		Blake2_128Concat,
+		Index,
+		MetadataExtensionDetails<T>,
+		OptionQuery,
+	>;
+
+	/// Records all token URIs associated with a given claimer who performed the metadata extension.
+	#[pallet::storage]
+	#[pallet::getter(fn claimer_token_uri)]
+	pub(super) type ClaimerTokenURI<T: Config> = StorageDoubleMap<
```

Comment:
should this be changed to `MetadataExtensions`? This is basically an `extension`, with different key-value format and not indexed.